neqo-bin/src/server/mod.rs

```diff
@@ -224,36 +225,57 @@ impl ServerRunner {
         }
     }
 
-    /// Tries to find a socket, but then just falls back to sending from the first.
-    fn find_socket(&mut self, addr: SocketAddr) -> &mut crate::udp::Socket {
-        let ((_host, first_socket), rest) = self.sockets.split_first_mut().unwrap();
-        rest.iter_mut()
-            .map(|(_host, socket)| socket)
-            .find(|socket| {
-                socket
-                    .local_addr()
-                    .ok()
-                    .map_or(false, |socket_addr| socket_addr == addr)
-            })
-            .unwrap_or(first_socket)
-    }
-
-    async fn process(&mut self, mut dgram: Option<Datagram>) -> Result<(), io::Error> {
+    async fn process(&mut self, mut ready_socket_index: Option<usize>) -> Result<(), io::Error> {
+        let mut input_dgrams: Option<DatagramIter> = None;
         loop {
-            match self.server.process(dgram.take(), (self.now)()) {
+            let input_dgram = if let Some(d) = input_dgrams.iter_mut().flatten().next() {
+                // Take datagram read in previous iteration.
+                Some(d)
+            } else {
+                // Try reading new datagrams from the socket.
+                if let Some(inx) = ready_socket_index {
+                    let (host, socket) = self.sockets.get_mut(inx).unwrap();
+                    input_dgrams = socket.recv(*host, &mut self.recv_buf)?;
+                }
+                // Then take the first datagram, if any.
+                input_dgrams.iter_mut().flatten().next().map_or_else(
+                    || {
+                        // Reading from the socket returned no datagrams. Don't try again.
+                        ready_socket_index = None;
+                        input_dgrams = None;
+                        None
+                    },
+                    Some,
+                )
```

Comment:
I'm surprised that clippy didn't pick this one up.

```suggestion
                input_dgrams.iter_mut().flatten().next().or_else(
                    || {
                        // Reading from the socket returned no datagrams. Don't try again.
                        ready_socket_index = None;
                        input_dgrams = None;
                        None
                    },
                )
```

Reply:
> I'm surprised that clippy didn't pick this one up.

And I am surprised that I didn't see this. :facepalm: Thank you Martin!

d23aa741